### PR TITLE
wasm-strip: Add `-k/--keep-section` flag to prevent sections with specific names from getting stripped

### PR DESCRIPTION
**Pull request by surma**
## TL;DR

This PR proposes the addition of the `-k/--keep-section <SECTION NAME>` flag, that can be specified multiple times to prevent sections with a specific name from getting stripped.

Below you’ll find some context why I think this is useful. WDYT?

## Context
This PR is motivated by my desire to strip everything _but_ the `name` section. I have found that for example Rust embeds a lot of other custom sections, which makes module size analysis quite noisy. Here, for example, the percentages become almost meaningless due to the size of the custom sections:

```
$ twiggy top bindgen.wasm -n 20
 Shallow Bytes │ Shallow % │ Item
───────────────┼───────────┼────────────────────────────────────────────────────────────────────────────
        652300 ┊    36.33% ┊ custom section '.debug_str'
        404594 ┊    22.53% ┊ custom section '.debug_info'
        285988 ┊    15.93% ┊ custom section '.debug_line'
        254962 ┊    14.20% ┊ custom section '.debug_pubnames'
        176000 ┊     9.80% ┊ custom section '.debug_ranges'
          4917 ┊     0.27% ┊ dlmalloc::dlmalloc::Dlmalloc<A>::malloc::h6c2f92ea78996c54
          4181 ┊     0.23% ┊ custom section '.debug_abbrev'
          3951 ┊     0.22% ┊ "function names" subsection
          1437 ┊     0.08% ┊ dlmalloc::dlmalloc::Dlmalloc<A>::free::h9921c2b76eae2534
          1129 ┊     0.06% ┊ __rdl_realloc
           725 ┊     0.04% ┊ dlmalloc::dlmalloc::Dlmalloc<A>::dispose_chunk::ha50595baecb6fca0
           553 ┊     0.03% ┊ __externref_table_alloc
           511 ┊     0.03% ┊ dlmalloc::Dlmalloc<A>::malloc::hd8efb8c25e917a05
           387 ┊     0.02% ┊ dlmalloc::dlmalloc::Dlmalloc<A>::release_unused_segments::he6aa823ad463a901
           352 ┊     0.02% ┊ dlmalloc::dlmalloc::Dlmalloc<A>::insert_large_chunk::hdf12b63cafdb133b
           346 ┊     0.02% ┊ dlmalloc::dlmalloc::Dlmalloc<A>::unlink_large_chunk::hb9b09f5b80b6743e
           324 ┊     0.02% ┊ custom section '.debug_pubtypes'
           319 ┊     0.02% ┊ compiler_builtins::mem::memcpy::h10893781353244d9
           307 ┊     0.02% ┊ __externref_table_dealloc
           259 ┊     0.01% ┊ __externref_heap_live_count
          2134 ┊     0.12% ┊ ... and 111 more.
       1795676 ┊   100.00% ┊ Σ [131 Total Rows]
```

However, `wasm-strip` removes the `names` section by default, which means that after stripping, I have no idea what is contributing to the size:

```
$ twiggy top bindgen.wasm -n 20
 Shallow Bytes │ Shallow % │ Item
───────────────┼───────────┼─────────────────────────────────────────────────────────────────────────
          4917 ┊    37.93% ┊ code[19]
          1437 ┊    11.08% ┊ code[24]
          1129 ┊     8.71% ┊ code[29]
           725 ┊     5.59% ┊ code[20]
           553 ┊     4.27% ┊ code[7]
           511 ┊     3.94% ┊ code[18]
           387 ┊     2.99% ┊ code[23]
           352 ┊     2.72% ┊ code[22]
           346 ┊     2.67% ┊ code[21]
           319 ┊     2.46% ┊ code[66]
           307 ┊     2.37% ┊ code[8]
           259 ┊     2.00% ┊ code[10]
           157 ┊     1.21% ┊ code[25]
            67 ┊     0.52% ┊ import __wbindgen_externref_xform__::__wbindgen_externref_table_set_null
            63 ┊     0.49% ┊ import __wbindgen_externref_xform__::__wbindgen_externref_table_grow
            63 ┊     0.49% ┊ code[14]
            59 ┊     0.46% ┊ code[59]
            58 ┊     0.45% ┊ code[1]
            49 ┊     0.38% ┊ code[9]
            48 ┊     0.37% ┊ code[16]
          1158 ┊     8.93% ┊ ... and 86 more.
         12964 ┊   100.00% ┊ Σ [106 Total Rows]
```

With this PR, I can keep the `name` section around (maybe also the `producers`), making the output useful again.

```
$ ./wasm-strip -k name -k producers bindgen.wasm
$ twiggy top bindgen.wasm -n 20
 Shallow Bytes │ Shallow % │ Item
───────────────┼───────────┼────────────────────────────────────────────────────────────────────────────
          4917 ┊    28.89% ┊ dlmalloc::dlmalloc::Dlmalloc<A>::malloc::h6c2f92ea78996c54
          3951 ┊    23.21% ┊ "function names" subsection
          1437 ┊     8.44% ┊ dlmalloc::dlmalloc::Dlmalloc<A>::free::h9921c2b76eae2534
          1129 ┊     6.63% ┊ __rdl_realloc
           725 ┊     4.26% ┊ dlmalloc::dlmalloc::Dlmalloc<A>::dispose_chunk::ha50595baecb6fca0
           553 ┊     3.25% ┊ __externref_table_alloc
           511 ┊     3.00% ┊ dlmalloc::Dlmalloc<A>::malloc::hd8efb8c25e917a05
           387 ┊     2.27% ┊ dlmalloc::dlmalloc::Dlmalloc<A>::release_unused_segments::he6aa823ad463a901
           352 ┊     2.07% ┊ dlmalloc::dlmalloc::Dlmalloc<A>::insert_large_chunk::hdf12b63cafdb133b
           346 ┊     2.03% ┊ dlmalloc::dlmalloc::Dlmalloc<A>::unlink_large_chunk::hb9b09f5b80b6743e
           319 ┊     1.87% ┊ compiler_builtins::mem::memcpy::h10893781353244d9
           307 ┊     1.80% ┊ __externref_table_dealloc
           259 ┊     1.52% ┊ __externref_heap_live_count
           157 ┊     0.92% ┊ dlmalloc::dlmalloc::Dlmalloc<A>::init_top::hd19f596040fbda30
            67 ┊     0.39% ┊ import __wbindgen_externref_xform__::__wbindgen_externref_table_set_null
            67 ┊     0.39% ┊ custom section 'producers'
            63 ┊     0.37% ┊ import __wbindgen_externref_xform__::__wbindgen_externref_table_grow
            63 ┊     0.37% ┊ __wbindgen_malloc
            59 ┊     0.35% ┊ <dlmalloc::sys::System as dlmalloc::Allocator>::alloc::hacc9295cf6d00ed7
            58 ┊     0.34% ┊ __wbindgen_describe_add
          1295 ┊     7.61% ┊ ... and 91 more.
         17022 ┊   100.00% ┊ Σ [111 Total Rows]
```

### CODE REVIEW

**Comment by sbc100** [comment]
While I think its fine/great to improve wabt in this way, I do wonder if it would be better for direct folks to the upstream llvm tools (llvm-strip, llvm-objcopy), which I think already come with both rust and emsdk.

Have you tried using the llvm tools to achieve your goals?

File: test/strip/keep_section.txt [hunk 1]
```diff
@@ -0,0 +1,34 @@
+;;; RUN: %(gen_wasm_py)s %(in_file)s -o %(temp_file)s.wasm
+;;; RUN: %(wasm-strip)s -k one %(temp_file)s.wasm -o %(temp_file)s_stripped.wasm
```
Comment (sbc100) [inline test/strip/keep_section.txt:2]: Maybe spell out the full option name here (`--keep-section`) so the reader can clearly see what is being tests?

File: src/tools/wasm-strip.cc [hunk 2]
```diff
@@ -76,9 +88,22 @@ class BinaryReaderStrip : public BinaryReaderNop {
     return stream_.WriteToFile(filename);
   }
 
+  Result BeginCustomSection(Index section_index,
+                            Offset size,
+                            std::string_view section_name) override {
+    if (sections_to_keep_.count(section_name) > 0) {
```
Comment (sbc100) [inline src/tools/wasm-strip.cc:94]: Isn't it ridiculous that we can't use `set::contains` until C++20?   I guess there must have been rationale at the time, but can you imagine designed a set API without a "contains/has" operation?

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
Can you prefix the PR title with `wasm-strip:`?

File: src/tools/wasm-strip.cc (hunk 2 above)
Comment (surma) [reply src/tools/wasm-strip.cc:94]: Ikr? I was really surprised by this as well

File: test/strip/keep_section.txt (hunk 1 above)
Comment (surma) [reply test/strip/keep_section.txt:2]:
```suggestion
;;; RUN: %(wasm-strip)s --keep-section one %(temp_file)s.wasm -o %(temp_file)s_stripped.wasm
```

**Comment by surma** [comment]
> Have you tried using the llvm tools to achieve your goals?

Actually @rreverser just educated me out-of-band that `llvm-strip` has `--keep-section` already. 

> would [it] be better for direct folks to the upstream llvm tools (llvm-strip, llvm-objcopy)?

If that's your suggestion (which I agree with, I just wasn't aware of these tools), should `wasm-strip` be deprecated?

Either way, I implemented the feedback, but I'll leave it up to you whether it's worth merging this.

**Comment by sbc100** [comment]
> If that's your suggestion (which I agree with, I just wasn't aware of these tools), should `wasm-strip` be deprecated?
> 

Maybe... I will bring this up in our test tools meeting.
